architecture-decisions/2024-06-06_arc4-by-default.md

```diff
@@ -0,0 +1,153 @@
+      return x + y;
+   }
+
+   abiAdd(x: uint64, y: uint64): uint64 {
+      return this.add(x, y);
+   }
+
+   @nonABIRouterFallback.call('NoOp')
+   nonAbiAdd(): void {
+      const x = btoi(this.txn.applicationArgs![0]);
+      const y = btoi(this.txn.applicationArgs![1]);
+      log(itob(this.add(x, y)));
+   }
+}
+```
+
+Handling of 'other' on completion actions can be done via decorators however the prescribed way in the documentation is to implement a [well-known](https://tealscript.netlify.app/guides/lifecycle/) method on your contract.
+
+Pros:
+ - A developer new to Algorand does not need to learn much about ARC4 in order to implement an ARC4 compliant contract.
+ - Flexibility exists for advanced devs to 'break out' of the ARC4 default when desired (kind of)
+
+Cons:
+ - The current implementation in TealScript outputs a lot of redundant teal in scenarios where an advanced user wants to do their own routing
+ - Doesn't give us a low friction path forward if ARC4 were to be replaced, or significantly updated
+ - In TypeScript, class members are public by default which could lead to accidentally exposing a method which shouldn't be.
+
+
+### Option 2 - Minimum Viable Smart Contract by default, ARC4 is opt in
+
+The approach taken by Alogrand Python is that a smart contract which extends the base `algopy.Contract` class will inherit no behaviour beyond that described in the context above (ie. `onComplete == clearState ? clearStateProgram() : approvalProgram()`).  This means the minimum viable smart contract (MVSC) would look something like this:
+
+```python
+# Contract which always approves
+class MVContract(Contract):
+    def approval_program(self) -> bool:
+        return True
+
+    def clear_state_program(self) -> bool:
+        return True
+```
+
+a TypeScript version might look like this:
+
+```ts
+export default class HelloWorldContract extends Contract {
+   approvalProgram(): boolean {
+      return true
+   }
+
+   clearState(): boolean {
+      return true
+   }
+}
+
+```
+
+An ARC4 compliant contract would instead extend the `algopy.arc4.ARC4Contract` base class. Each method must be decorated with either `@abimethod` or `@subroutine` to indicate a public or private method. 
+
+```python
+class HelloWorldContract(ARC4Contract):
+    @arc4.abimethod
+    def hello(self, name: String) -> String:
+        return "Hello, " + name
+```
+
+or in TypeScript
+
+```ts
+export default class HelloWorldContract extends arc4.Arc4Contract {
+  @arc4.abimethod()
+  sayHello(name: str): arc4.Str {
+    return new arc4.Str(`Hello ${name}`)
+  }
+}
+```
+
+Pros:
+- ARC4 support is built on top of a framework that can function in its own right giving us greater confidence we can support the next ARC if/when it comes, without breaking existing ARC4 support
+- MVSC gives power users absolute control without ARC4 getting in the way
+- Explicit decorators reduce the likelihood of accidentally exposing a method publicly.
```

Comment:
Want to note that decorators were initially how ARC4 methods worked in TEALScript, but I got feedback from TypeScript developers that frequent use of decorators felt strange to them.  Also worth noting that they are still considered experimental in TypeScript and not enabled by default with most configurations.

Reply:
Decorators are stage 3 in TC39, TypeScript's experimental decorators refers to the capturing of typescript meta data which differs to the TC39 proposal. We shouldn't need to enable typescript's experimental decorators.

Reply:
Interesting... I know TEALScript contracts fail to pass TypeScript diagnostics without this flag set if they use decorators. I think we should look into this a bit more and make sure we reflect it properly in the ADR

Reply:
It might be you're using the experimental decorator method signature.  It's working fine [here](https://www.typescriptlang.org/play/?#code/GYVwdgxgLglg9mABAEwKYTgJwIZSwCim0wHNUoAuRbMATwBpEBbc7ZXbKm2gSkQG8AUIkQYwAZzgAbVADopcEoWJkojFkXZEeg3QF9doSLATUpUNBhx5M+OAAcTEqgCV0WZAB5xUTDDAkjNwAfHxCwoiY5CCYSEbQ8EjKpORcdOqsWpzUdGERImKSMvKKyaoZmhyMDk7iOroiBga6EFLY4uKIALK0AMJtHQIRAAKWWLhYI9jmY9YE-NRUAIyMAEZUAEyIejoiAMRQqD74eRFNQA) with experimentalDecorators disabled

Reply:
Ah ok yep, it is because of the signature of the decorator.